#' @references 
#' 
#' Kanehisa, M., Furumichi, M., Tanabe, 
#' M., Sato, Y., & Morishima, K. (2017). 
#' KEGG: new perspectives on genomes, pathways, diseases and drugs. 
#' Nucleic acids research, 45(D1), D353-D361.
#' 
#' Karnovsky, A., Weymouth, T., Hull, T., Tarcea, V. G., 
#' Scardoni, G., Laudanna, C., ... & Athey, B. (2011). 
#' Metscape 2 bioinformatics tool for the analysis 
#' and visualization of metabolomics and gene expression data. 
#' Bioinformatics, 28(3), 373-380.
#' 
#' Tenenbaum, D. (2013). KEGGREST: Client-side REST access 
#' to KEGG. R package version, 1(1).
#' 
#' Chang, W., Cheng, J., Allaire, JJ., 
#' Xie, Y., & McPherson, J. (2017).
#' shiny: Web Application Framework for R. R package version 1.0.5.
#' https://CRAN.R-project.org/package=shiny
#' 
#' Picart-Armada, S., Fernandez-Albert, F., Vinaixa, 
#' M., Rodriguez, M. A., Aivio, S., Stracker, 
#' T. H., Yanes, O., & Perera-Lluna, A. (2017). 
#' Null diffusion-based enrichment for metabolomics data. 
#' PLOS ONE, 12(12), e0189012.
